File: src/randombytes.c
#include <assert.h>
#include <stdio.h>

typedef unsigned char u8;
typedef unsigned long long u64;

void randombytes(u8 *buf, u64 nbytes) {
    FILE *fp = 0; 

    assert(buf && nbytes);
    assert(fp = fopen("/dev/urandom", "r"));
    assert((fread(buf, nbytes, 1, fp)) == 1);
    assert(!fclose(fp));
}

File: src/randombytes.rs
use rand::{OsRng, Rng};
use std::slice;

#[no_mangle]
pub unsafe extern fn randombytes(buf: *mut u8, nbytes: u64) {
    assert!(!buf.is_null());
    let mut rng = OsRng::new().ok().expect("OSRNG failed.");
    let mut bytes = slice::from_raw_parts_mut(buf, nbytes as usize);
    rng.fill_bytes(bytes);
}


